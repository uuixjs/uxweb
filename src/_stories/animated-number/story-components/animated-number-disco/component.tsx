import * as React from "react";

import { AnimatedNumber, Layout } from "v2";

import { Component } from "react";

interface DiscoProps {
  startValue: number;
  deltaMin?: number;
  deltaMax?: number;
}

interface DiscoState {
  value: number;
}

export class AnimatedNumberDisco extends Component<DiscoProps, DiscoState> {
  private interval: number | undefined;

  public constructor(props: DiscoProps) {
    super(props);
    this.state = {
      value: props.startValue,
    };
  }

  public componentDidMount() {
    this.interval = window.setInterval(this.tick, 2000);
  }

  public componentWillUnmount() {
    window.clearInterval(this.interval);
  }

  public render() {
    return (
      <Layout>
        <AnimatedNumber
          value={this.state.value}
          formatValue={(n) => n.toLocaleString()}
        />
      </Layout>
    );
  }

  private tick = () => {
    const diff = generateRandomInteger(
      this.props.deltaMin || -10,
      this.props.deltaMax || 10,
    );

    this.setState((prevState) => ({
      value: prevState.value + diff,
    }));
  };
}

function generateRandomInteger(min: number, max: number) {
  return Math.floor(min + Math.random() * (max + 1 - min));
}
