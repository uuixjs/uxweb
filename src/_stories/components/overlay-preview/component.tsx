import "./styles.scss";

import * as React from "react";

import {
  AlignItems,
  Background,
  CoreImage,
  Display,
  InjectLayout,
  Layout,
  Overflow,
  OverlayRegion,
  Position,
  Radio,
  TransitionGroup,
  TransitionType,
  ZIndex,
} from "v2";
import { Component, FormEvent } from "react";
import { Padding, cn, newUUIDv4 } from "@uuixjs/uuixweb-lib";

import { ExampleThumbnails } from "../../assets";

export enum ExampleBG {
  Off = "non-overlay",
  Photo1 = "photo-1",
  Photo2 = "photo-2",
  Video = "video",
  Black = "black",
  White = "white",
}

interface Props {
  defaultBg?: ExampleBG;
  blur?: boolean;
  padding?: Padding;
}

interface State {
  bg: ExampleBG;
}

export class OverlayPreview extends Component<Props, State> {
  public static defaultProps = {
    padding: 2,
    blur: true,
  };

  private uid = newUUIDv4();

  public constructor(props: Props) {
    super(props);
    this.state = {
      bg: props.defaultBg || ExampleBG.Black,
    };
  }

  public render() {
    const overlay = this.state.bg !== ExampleBG.Off;

    return (
      <Layout>
        <Layout display={Display.Flex} padding={1}>
          {Object.keys(ExampleBG).map((k, i) => (
            <Layout key={k} padding={{ left: i > 0 ? 1 : 0 }}>
              <Radio
                name={`overlay-preview-${this.uid}`}
                label={k}
                value={ExampleBG[k as keyof typeof ExampleBG]}
                onChange={this.onRadioSelected}
                checked={
                  this.state.bg === ExampleBG[k as keyof typeof ExampleBG]
                }
              />
            </Layout>
          ))}
        </Layout>

        <OverlayRegion overlay={overlay}>
          <Layout
            className="overlay-preview__container"
            position={Position.Relative}
            background={overlay ? Background.Overlay : undefined}
            padding={this.props.padding}
          >
            {this.renderBackground()}
            <Layout zIndex={ZIndex.Above} position={Position.Relative}>
              {this.props.children}
            </Layout>
          </Layout>
        </OverlayRegion>
      </Layout>
    );
  }

  private renderBackground() {
    let element;
    if (this.state.bg === ExampleBG.Photo1) {
      element = <CoreImage src={ExampleThumbnails.stream1} alt="" />;
    }

    if (this.state.bg === ExampleBG.Photo2) {
      element = <CoreImage src={ExampleThumbnails.stream2} alt="" />;
    }

    if (this.state.bg === ExampleBG.Video) {
      element = (
        <video
          src="https://d3hmvhl7ru3t12.cloudfront.net/video/pages/home/header-82b9abce067992beedde6d103b6241d616a7ee97cc0d4f3126277098c2e4b09004325f25c6a245656096bd333c880b0032e2cbdbed9d2514220e20718482e95b.mp4"
          autoPlay
          muted
          loop
        />
      );
    }

    if (this.state.bg === ExampleBG.Black) {
      element = <div style={{ background: "#000" }} />;
    }

    if (this.state.bg === ExampleBG.White) {
      element = <div style={{ background: "#FFF" }} />;
    }

    return (
      <InjectLayout
        display={Display.Flex}
        alignItems={AlignItems.Stretch}
        attachTop
        attachLeft
        attachBottom
        attachRight
        overflow={Overflow.Hidden}
      >
        <TransitionGroup
          className="overlay-preview__transition-group-container"
          transitionType={TransitionType.Fade}
        >
          {element ? (
            <InjectLayout
              className={cn({
                ["overlay-preview__bg-element"]: true,
                ["overlay-preview__bg-element--blur"]: this.props.blur,
              })}
              key={this.state.bg}
              position={Position.Absolute}
              attachTop
              attachLeft
              attachBottom
              attachRight
              fullWidth
              fullHeight
            >
              {element}
            </InjectLayout>
          ) : (
            []
          )}
        </TransitionGroup>
      </InjectLayout>
    );
  }

  private onRadioSelected = (e: FormEvent<HTMLInputElement>) => {
    if (e.currentTarget.checked) {
      this.setState({
        bg: e.currentTarget.value as ExampleBG,
      });
    }
  };
}
