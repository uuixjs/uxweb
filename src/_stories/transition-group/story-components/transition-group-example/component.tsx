import * as React from "react";

import {
  AlignItems,
  Background,
  Button,
  ButtonType,
  Column,
  Display,
  FlexDirection,
  FlexWrap,
  FormGroup,
  Grid,
  Input,
  InputType,
  JustifyContent,
  Layout,
  Range,
  Select,
  Tower,
  TowerChildWidth,
  TowerGutter,
  TransitionDuration,
  TransitionGroup,
  TransitionType,
  TransitionTypeOption,
} from "v2";
import { Component, FormEvent, MouseEvent } from "react";

import { TransitionGroupChecklist } from "../transition-group-checklist";
import { newUUIDv4 } from "@uuixjs/uuixweb-lib";

enum DemoLayoutType {
  Ordinary = "ordinary",
  Flexbox = "flexbox",
  // eslint-disable-next-line @typescript-eslint/no-shadow
  Tower = "tower",
}

interface Item {
  key: string;
}

export interface UpdateSettings {
  numToAdd: number;
  addAtIndex: number;
  numToRemove: number;
  removeAtIndex: number;
  numToMove: number;
  moveFromIndex: number;
  moveToIndex: number;
}

export interface LayoutSettings {
  layout: DemoLayoutType;
  direction: FlexDirection;
  duration: TransitionDuration | number;
  transitionType: TransitionTypeOption;
  boxSize: number;
}

interface Props {}

interface State extends UpdateSettings, LayoutSettings {
  showItems: boolean;
  items: Item[];
}

const INITIAL_VALUES: UpdateSettings = {
  numToAdd: 1,
  addAtIndex: 1,
  numToRemove: 0,
  removeAtIndex: 1,
  numToMove: 0,
  moveFromIndex: 0,
  moveToIndex: 1,
};

export class TransitionGroupExample extends Component<Props, State> {
  public state: State = {
    ...INITIAL_VALUES,
    layout: DemoLayoutType.Ordinary,
    direction: FlexDirection.Column,
    duration: TransitionDuration.Long,
    transitionType: [TransitionType.ScaleOver, TransitionType.ScaleOver],
    boxSize: 240,
    showItems: true,
    items: [],
  };

  public componentDidMount() {
    this.init();
  }

  public render() {
    return (
      <Layout margin={{ top: 2 }} border background={Background.Base}>
        <Grid>
          <Column cols={{ default: 12, md: 6, lg: 5, xxl: 4 }}>
            {this.renderListControls()}
          </Column>

          <Column cols={{ default: 12, md: 6, lg: 7, xxl: 8 }}>
            {this.renderLayoutControls()}
          </Column>

          <Column cols={{ default: 12, lg: 12, xxl: 8 }}>
            {this.state.showItems && (
              <div
                style={{
                  display: "block",
                  border: "2px dashed lightgray",
                  margin: "2rem",
                }}
              >
                {this.renderExample()}
              </div>
            )}
          </Column>

          <Column cols={{ default: 12, lg: 12, xxl: 4 }}>
            <Layout padding={2}>
              <TransitionGroupChecklist
                setValues={this.setValues}
                updateView={this.changeItems}
                setItemCount={this.resetItemsWithoutTransition}
                addItem={this.addItem}
                removeItem={this.removeItem}
              />
            </Layout>
          </Column>
        </Grid>
      </Layout>
    );
  }

  private renderExample() {
    if (this.state.layout === DemoLayoutType.Ordinary) {
      return (
        <TransitionGroup
          duration={this.state.duration}
          transitionType={this.state.transitionType}
        >
          {this.renderItems()}
        </TransitionGroup>
      );
    }
    if (this.state.layout === DemoLayoutType.Flexbox) {
      return (
        <TransitionGroup
          duration={this.state.duration}
          transitionType={this.state.transitionType}
          component={
            <Layout
              display={Display.Flex}
              flexDirection={this.state.direction}
              flexWrap={FlexWrap.Wrap}
              padding={1}
            />
          }
        >
          {this.renderItems()}
        </TransitionGroup>
      );
    }

    if (this.state.layout === DemoLayoutType.Tower) {
      return (
        <TransitionGroup
          duration={this.state.duration}
          transitionType={this.state.transitionType}
          component={
            <Tower
              childWidth={TowerChildWidth.Medium}
              gutterSize={TowerGutter.ExtraSmall}
              placeholderItems={20}
            />
          }
        >
          {this.renderItems()}
        </TransitionGroup>
      );
    }
  }

  private renderItems() {
    return this.state.items.map((item, index) => (
      <div
        key={item.key}
        style={{
          display: "flex",
          flexDirection: "column",
          justifyContent: "center",
          alignItems: "center",
          width: `100%`,
          height: this.state.boxSize ? `${this.state.boxSize / 2}px` : "100px",
          background: "green",
          color: "rgba(255,255,255,0.9)",
          fontSize: "1.5rem",
          border: "2px solid darkgreen",
          margin: "",
        }}
      >
        <Layout
          display={Display.Flex}
          flexDirection={FlexDirection.Column}
          alignItems={AlignItems.Center}
          margin={{ bottom: 1 }}
        >
          <div>index: {index}</div>
          <Layout display={Display.Flex} alignItems={AlignItems.Center}>
            <div>key: {item.key.substring(item.key.length - 6)}</div>
            <div
              style={{
                width: "20px",
                height: "20px",
                background: `#${item.key.substring(item.key.length - 6)}`,
                margin: "0 0.5rem",
              }}
            />
          </Layout>
        </Layout>
        <Button
          variant={ButtonType.Secondary}
          onClick={() => this.removeItem(item.key)}
          overlay
        >
          Remove
        </Button>
      </div>
    ));
  }

  private renderListControls = () => {
    return (
      <form onSubmit={this.onSubmitForm}>
        <Layout
          display={Display.InlineFlex}
          flexDirection={FlexDirection.Column}
          alignItems={AlignItems.End}
          padding={2}
        >
          <Layout margin={{ bottom: 1 }}>
            Move
            <div
              style={{
                display: "inline-block",
                width: "50px",
                margin: "0 0.5rem",
              }}
            >
              <Input
                type={InputType.Number}
                value={this.state.numToMove.toString()}
                onChange={this.setNumToMove}
                min={0}
                max={10}
              />
            </div>
            items from
            <div
              style={{
                display: "inline-block",
                width: "50px",
                margin: "0 0.5rem",
              }}
            >
              <Input
                type={InputType.Number}
                value={this.state.moveFromIndex.toString()}
                onChange={this.setMoveFromIndex}
                min={0}
              />
            </div>
            to
            <div
              style={{
                display: "inline-block",
                width: "50px",
                margin: "0 0 0 0.5rem",
              }}
            >
              <Input
                type={InputType.Number}
                value={this.state.moveToIndex.toString()}
                onChange={this.setMoveToIndex}
                min={0}
              />
            </div>
          </Layout>

          <Layout margin={{ top: 0.5 }}>
            Add
            <div
              style={{
                display: "inline-block",
                width: "50px",
                margin: "0 0.5rem",
              }}
            >
              <Input
                type={InputType.Number}
                value={this.state.numToAdd.toString()}
                onChange={this.setNumToAdd}
                min={0}
                max={10}
              />
            </div>
            items at index
            <div
              style={{
                display: "inline-block",
                width: "50px",
                margin: "0 0 0 0.5rem",
              }}
            >
              <Input
                type={InputType.Number}
                value={this.state.addAtIndex.toString()}
                onChange={this.setAddAtIndex}
                min={0}
              />
            </div>
          </Layout>

          <Layout margin={{ top: 0.5 }}>
            Remove
            <div
              style={{
                display: "inline-block",
                width: "50px",
                margin: "0 0.5rem",
              }}
            >
              <Input
                type={InputType.Number}
                value={this.state.numToRemove.toString()}
                onChange={this.setNumToRemove}
                min={0}
                max={10}
              />
            </div>
            items at index
            <div
              style={{
                display: "inline-block",
                width: "50px",
                margin: "0 0 0 0.5rem",
              }}
            >
              <Input
                type={InputType.Number}
                value={this.state.removeAtIndex.toString()}
                onChange={this.setRemoveAtIndex}
                min={0}
              />
            </div>
          </Layout>

          <Layout display={Display.Flex} margin={{ top: 1 }}>
            <Layout margin={{ right: 1 }}>
              <Button onClick={this.onSubmitForm}>Update</Button>
            </Layout>
            <Button onClick={this.init} variant={ButtonType.Secondary}>
              Reset
            </Button>
          </Layout>
        </Layout>
      </form>
    );
  };

  private generateItem = (): Item => {
    return {
      key: newUUIDv4(),
    };
  };

  private init = (e?: MouseEvent<HTMLElement>) => {
    if (e) {
      e.preventDefault();
    }

    this.setState({
      ...INITIAL_VALUES,
      // Initial items have static keys and color values to ensure screenshot diff passes
      items: [{ key: "aa0000" }, { key: "bbbb00" }, { key: "0033cc" }],
    });
  };

  private resetItemsWithoutTransition = async (num: number) => {
    return new Promise<void>((resolve) => {
      this.setState(
        {
          showItems: false,
        },
        () => {
          // wait for full re-render
          requestAnimationFrame(() => {
            this.setState(
              {
                showItems: true,
                items: Array.from(Array(num), this.generateItem),
              },
              resolve,
            );
          });
        },
      );
    });
  };

  private onSubmitForm = (
    e: FormEvent<HTMLFormElement> | MouseEvent<HTMLElement>,
  ) => {
    e.preventDefault();
    this.changeItems();
  };

  private changeItems = () => {
    this.setState((prevState) => {
      const {
        numToAdd,
        addAtIndex,
        numToRemove,
        removeAtIndex,
        numToMove,
        moveFromIndex,
        moveToIndex,
      } = prevState;

      const items = prevState.items.concat();
      const newItems = Array.from(Array(numToAdd), this.generateItem);

      // Remove some
      items.splice(removeAtIndex, numToRemove);

      // Add some
      items.splice(addAtIndex, 0, ...newItems);

      // Move some
      const moving = items.splice(moveFromIndex, numToMove);
      items.splice(moveToIndex, 0, ...moving);

      return {
        items,
      };
    });
  };

  private addItem = (key: string, index: number = 0) => {
    this.setState((prevState) => {
      const items = prevState.items.concat();
      items.splice(index, 0, { key });
      return {
        items,
      };
    });
  };

  private removeItem = (key: string) => {
    this.setState((prevState) => ({
      items: prevState.items.filter((item) => item.key !== key),
    }));
  };

  private setNumToAdd = (e: FormEvent<HTMLInputElement>) => {
    this.setState({
      numToAdd: parseInt(e.currentTarget.value, undefined),
    });
  };

  private setAddAtIndex = (e: FormEvent<HTMLInputElement>) => {
    this.setState({
      addAtIndex: parseInt(e.currentTarget.value, undefined),
    });
  };

  private setNumToRemove = (e: FormEvent<HTMLInputElement>) => {
    this.setState({
      numToRemove: parseInt(e.currentTarget.value, undefined),
    });
  };

  private setRemoveAtIndex = (e: FormEvent<HTMLInputElement>) => {
    this.setState({
      removeAtIndex: parseInt(e.currentTarget.value, undefined),
    });
  };

  private setNumToMove = (e: FormEvent<HTMLInputElement>) => {
    this.setState({
      numToMove: parseInt(e.currentTarget.value, undefined),
    });
  };

  private setMoveFromIndex = (e: FormEvent<HTMLInputElement>) => {
    this.setState({
      moveFromIndex: parseInt(e.currentTarget.value, undefined),
    });
  };

  private setMoveToIndex = (e: FormEvent<HTMLInputElement>) => {
    this.setState({
      moveToIndex: parseInt(e.currentTarget.value, undefined),
    });
  };

  private setValues = async (settings?: Partial<UpdateSettings>) => {
    return new Promise<void>((resolve) =>
      this.setState(
        {
          ...INITIAL_VALUES,
          ...settings,
        },
        resolve,
      ),
    );
  };

  private renderLayoutControls() {
    return (
      <Layout
        padding={2}
        fullWidth
        display={Display.Flex}
        justifyContent={JustifyContent.Start}
        flexWrap={FlexWrap.Wrap}
      >
        <Layout flexGrow={0} margin={{ right: 2 }}>
          <FormGroup label="Layout Demo">
            <Select value={this.state.layout} onChange={this.updateLayout}>
              {Object.keys(DemoLayoutType).map((key: string) => (
                <option
                  value={DemoLayoutType[key as keyof typeof DemoLayoutType]}
                  key={DemoLayoutType[key as keyof typeof DemoLayoutType]}
                >
                  {key}
                </option>
              ))}
            </Select>
          </FormGroup>
        </Layout>

        <Layout flexGrow={0} margin={{ right: 2 }}>
          <FormGroup label="FlexDirection">
            <Select
              value={this.state.direction}
              onChange={this.updateDirection}
              disabled
            >
              <option value={FlexDirection.Row}>Row</option>
              <option value={FlexDirection.Column}>Column</option>
            </Select>
          </FormGroup>
        </Layout>

        <Layout flexGrow={0} margin={{ right: 2 }}>
          <FormGroup label="TransitionDuration">
            <Select
              value={String(this.state.duration)}
              onChange={this.updateDuration}
            >
              {Object.keys(TransitionDuration).map((key: string) => (
                <option
                  value={
                    TransitionDuration[key as keyof typeof TransitionDuration]
                  }
                  key={
                    TransitionDuration[key as keyof typeof TransitionDuration]
                  }
                >
                  {key}
                </option>
              ))}
              <option value="1000" key="1000ms">
                1,000ms
              </option>
              <option value="3000" key="3000ms">
                3,000ms
              </option>
              <option value="5000" key="5000ms">
                5,000ms
              </option>
              <option value="10000" key="10000ms">
                10,000ms
              </option>
              <option value="0" key="0ms">
                0ms
              </option>
            </Select>
          </FormGroup>
        </Layout>

        <Layout flexGrow={0} margin={{ right: 2 }}>
          <FormGroup label="TransitionType [Enter]">
            <Select
              value={this.state.transitionType[0]}
              onChange={this.updateTransitionTypeEnter}
            >
              {Object.keys(TransitionType).map((key: string) => (
                <option
                  value={TransitionType[key as keyof typeof TransitionType]}
                  key={TransitionType[key as keyof typeof TransitionType]}
                >
                  {key}
                </option>
              ))}
            </Select>
          </FormGroup>
        </Layout>

        <Layout flexGrow={0} margin={{ right: 2 }}>
          <FormGroup label="TransitionType [Exit]">
            <Select
              value={this.state.transitionType[1]}
              onChange={this.updateTransitionTypeExit}
            >
              {Object.keys(TransitionType).map((key: string) => (
                <option
                  value={TransitionType[key as keyof typeof TransitionType]}
                  key={TransitionType[key as keyof typeof TransitionType]}
                >
                  {key}
                </option>
              ))}
            </Select>
          </FormGroup>
        </Layout>

        <Layout flexGrow={0} margin={{ right: 2 }}>
          <FormGroup label="Box Size">
            <Range
              min={100}
              max={600}
              value={this.state.boxSize.toString()}
              onChange={this.updateBoxSize}
              disabled
            />
          </FormGroup>
        </Layout>
      </Layout>
    );
  }

  private updateLayout = (e: FormEvent<HTMLSelectElement>) => {
    this.setState({
      layout: e.currentTarget.value as DemoLayoutType,
    });
  };

  private updateDirection = (e: FormEvent<HTMLSelectElement>) => {
    this.setState({
      direction: e.currentTarget.value as FlexDirection,
    });
  };

  private updateDuration = (e: FormEvent<HTMLSelectElement>) => {
    if (e.currentTarget.value.match(/^[0-9]+$/)) {
      this.setState({
        duration: Number(e.currentTarget.value),
      });
    } else {
      this.setState({
        duration: e.currentTarget.value as TransitionDuration,
      });
    }
  };

  private updateTransitionTypeEnter = (e: FormEvent<HTMLSelectElement>) => {
    this.setState({
      transitionType: [
        e.currentTarget.value as TransitionType,
        this.state.transitionType[1] as TransitionType,
      ],
    });
  };

  private updateTransitionTypeExit = (e: FormEvent<HTMLSelectElement>) => {
    this.setState({
      transitionType: [
        this.state.transitionType[0] as TransitionType,
        e.currentTarget.value as TransitionType,
      ],
    });
  };

  private updateBoxSize = (e: FormEvent<HTMLInputElement>) => {
    this.setState({
      boxSize: Number.parseInt(e.currentTarget.value, undefined),
    });
  };
}
