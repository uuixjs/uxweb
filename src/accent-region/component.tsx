import {
  AccentToken,
  AccentTokenValues,
  accentTokenDarkThemeMap,
  accentTokenLightThemeMap,
  formatAccentTokenCssVars,
  formatTokenOverrides,
} from "./token-map";
import { RGBColor, parse, rgb2hex } from "lib/pure-color";
import {
  ThemeContextValue,
  accentColorGenerator,
  colorIsDark,
  formatColor,
  getHoverColor,
  styled,
  useTheme,
} from "@uuixjs/uuixweb-lib";

import { FC } from "react";
import { StaticTokenMap } from "lib/ui-tokens";
import { TokenOverrideRegion } from "../token-override-region";

export interface AccentRegionProps {
  /** The original chosen color */
  inputColor?: string | null;
  /** Set to true if the contrast ratio between the original color and `#FFFFFF` is less than 4.5 */
  inputColorIsDark?: boolean | null;
  /** Color generated according to hover color spec */
  inputColorHover?: string | null;
  /** Color generated according to `accentColorGenerator` spec */
  primary1?: string | null;
  /** Color generated according to `accentColorGenerator` spec */
  primary2?: string | null;
  /** Color generated according to `accentColorGenerator` spec */
  primary3?: string | null;
  /** Color generated according to `accentColorGenerator` spec */
  primary4?: string | null;
  /** Color generated according to `accentColorGenerator` spec */
  primary5?: string | null;
}

export const ScAccentRegionCssVars = styled.div<{ $values: AccentTokenValues }>`
  /** Accent Token CSS Vars: */
  ${({ $values }) => formatAccentTokenCssVars($values)}
`;

/**
 * Changes the color palette for a region of the page
 * and should be populated with values which have been generated
 * according to the Core UI accent color generator spec.
 *
 * Values may be generated and stored on the server, or can be generated
 * at render time by using the functions provided in the Core UI package.
 */
export const AccentRegion: FC<AccentRegionProps> = ({
  inputColor,
  inputColorIsDark,
  inputColorHover,
  primary1,
  primary2,
  primary3,
  primary4,
  primary5,
  children,
}) => {
  const tokenValues: AccentTokenValues = {};

  if (inputColor) {
    tokenValues[AccentToken.Accent] = formatColor(inputColor);
  }

  if (typeof inputColorIsDark === "boolean") {
    tokenValues[AccentToken.AccentLabel] = inputColorIsDark
      ? "#FFFFFF"
      : "#000000";
  }

  if (inputColorHover) {
    tokenValues[AccentToken.Hover] = formatColor(inputColorHover);
  }

  if (typeof inputColorIsDark === "boolean") {
    tokenValues[AccentToken.BackgroundButtonPurchase] = inputColorIsDark
      ? StaticTokenMap["color-opac-b-5"]
      : StaticTokenMap["color-opac-w-5"];
  }

  if (primary1) {
    tokenValues[AccentToken.Primary1] = formatColor(primary1);
  }

  if (primary2) {
    tokenValues[AccentToken.Primary2] = formatColor(primary2);
  }

  if (primary3) {
    tokenValues[AccentToken.Primary3] = formatColor(primary3);
  }

  if (primary4) {
    tokenValues[AccentToken.Primary4] = formatColor(primary4);
  }

  if (primary5) {
    tokenValues[AccentToken.Primary5] = formatColor(primary5);
  }

  const theme = useTheme();
  let tokenOverrides: ThemeContextValue["tokenOverrides"];

  if (theme === undefined || theme.name === "dark") {
    tokenOverrides = formatTokenOverrides(tokenValues, accentTokenDarkThemeMap);
  } else if (theme.name === "light") {
    tokenOverrides = formatTokenOverrides(
      tokenValues,
      accentTokenLightThemeMap,
    );
  }

  return (
    <TokenOverrideRegion tokenOverrides={tokenOverrides}>
      <ScAccentRegionCssVars $values={tokenValues}>
        {children}
      </ScAccentRegionCssVars>
    </TokenOverrideRegion>
  );
};

export function generateAccentRegionProps(
  color: string | null,
): AccentRegionProps {
  const rgb = color && (parse(formatColor(color)) as RGBColor);

  if (!rgb) {
    return {
      inputColor: null,
      inputColorIsDark: false,
      inputColorHover: null,
      primary1: null,
      primary2: null,
      primary3: null,
      primary4: null,
      primary5: null,
    };
  }

  const colors = accentColorGenerator(rgb).map((v) => rgb2hex(v));

  return {
    inputColor: rgb2hex(rgb),
    inputColorIsDark: colorIsDark(rgb),
    inputColorHover: rgb2hex(getHoverColor(rgb)),
    primary1: colors[0],
    primary2: colors[1],
    primary3: colors[2],
    primary4: colors[3],
    primary5: colors[4],
  };
}
