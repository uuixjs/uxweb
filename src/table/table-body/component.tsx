import { FC, HTMLProps } from "react";

import { cn } from "lib";

export const TableBody: FC<HTMLProps<HTMLTableSectionElement>> = (props) => {
  return <tbody {...props} className={cn("tw-table-body", props.className)} />;
};

TableBody.displayName = "TableBody";
