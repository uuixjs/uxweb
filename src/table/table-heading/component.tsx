import {
  AlignItems,
  Color,
  Display,
  Layout,
  TextAlign,
  VerticalAlign,
} from "../../layout";
import { FC, HTMLProps, MouseEventHandler, ReactNode } from "react";
import {
  Padding,
  staticTokenRule,
  styleVariant,
  styled,
  themeTokenRule,
} from "lib";

import { CoreText } from "../../core-text";
import { Icon } from "../../icon";
import { Interactable } from "../../interactable";
import { SVGAsset } from "../../svg";

export enum Sorting {
  Default = "default",
  Ascending = "ascending",
  Descending = "descending",
}

export interface TableHeadingProps extends ScTableHeadingProps {
  label?: string;
  children?: ReactNode;
  onClick?: MouseEventHandler<HTMLElement>;
  padding?: Padding;
  sorting?: Sorting | "default" | "ascending" | "descending"; // TODO: https://jira.xarth.tv/browse/COREUI-3554;
}

interface ScTableHeadingProps {
  textAlign?: TextAlign;
  verticalAlign?: VerticalAlign;
}

const ScTableHeading = styled.th<ScTableHeadingProps>`
  border-bottom-width: ${staticTokenRule("border-width-default")};
  border-bottom-style: solid;
  border-bottom-color: ${themeTokenRule("color-border-base")};
  background-color: ${themeTokenRule("color-background-alt")};

  vertical-align: ${styleVariant("verticalAlign", {
    [VerticalAlign.Top]: "top",
    [VerticalAlign.Middle]: "middle",
    [VerticalAlign.Bottom]: "bottom",
    [VerticalAlign.TextTop]: "text-top",
    [VerticalAlign.TextBottom]: "text-bottom",
    [VerticalAlign.Baseline]: "baseline",
  })};

  text-align: ${styleVariant("textAlign", {
    [TextAlign.Left]: "left",
    [TextAlign.Center]: "center",
    [TextAlign.Right]: "right",
    [TextAlign.Justify]: "justify",
  })};

  font-weight: ${staticTokenRule("font-weight-semibold")};

  &:first-child {
    border-top-left-radius: ${staticTokenRule("border-radius-medium")};
  }

  &:last-child {
    border-top-right-radius: ${staticTokenRule("border-radius-medium")};
  }
`;

export const TableHeading: FC<
  TableHeadingProps & Omit<HTMLProps<HTMLTableHeaderCellElement>, "ref" | "as">
> = ({
  label,
  textAlign = TextAlign.Left,
  padding = 1,
  sorting,
  verticalAlign = VerticalAlign.Middle,
  onClick,
  children,
  ...thProps
}) => {
  let heading: JSX.Element | undefined;
  let sortArrow: JSX.Element | undefined;

  const content = (
    <>
      {label && <CoreText bold>{label}</CoreText>}
      {children}
    </>
  );

  if (sorting !== undefined) {
    switch (sorting) {
      case "ascending":
        sortArrow = <Icon asset={SVGAsset.GlyphArrUp} />;
        break;

      case "descending":
        sortArrow = <Icon asset={SVGAsset.GlyphArrDown} />;
        break;
      case "default":
      default:
        sortArrow = <Icon asset={SVGAsset.GlyphArrUpDown} />;
        break;
    }

    heading = (
      <Interactable onClick={onClick}>
        <Layout
          color={Color.Link}
          display={Display.Flex}
          alignItems={AlignItems.Center}
          padding={padding}
        >
          <Layout margin={{ right: 0.5 }} flexGrow={1}>
            {content}
          </Layout>
          {sortArrow}
        </Layout>
      </Interactable>
    );
  } else {
    heading = <Layout padding={padding}>{content}</Layout>;
  }

  return (
    <ScTableHeading
      textAlign={textAlign}
      verticalAlign={verticalAlign}
      {...thProps}
      className="tw-table-heading"
    >
      {heading}
    </ScTableHeading>
  );
};

TableHeading.displayName = "TableHeading";
