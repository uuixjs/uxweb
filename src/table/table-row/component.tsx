import { FC, HTMLProps } from "react";

import { cn } from "@uuixjs/uuixweb-lib";

export const TableRow: FC<HTMLProps<HTMLTableRowElement>> = (props) => {
  return <tr {...props} className={cn("tw-table-row", props.className)} />;
};

TableRow.displayName = "TableRow";
