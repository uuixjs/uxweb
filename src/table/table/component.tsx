import { staticTokenRule, styled, themeTokenRule } from "lib";

export interface TableProps {
  alternateRows?: boolean;
}
export const Table = styled.table.attrs(() => ({
  className: "tw-table",
}))<TableProps>`
  width: 100%;
  border-radius: ${staticTokenRule("border-radius-medium")};
  box-shadow: 0 0 0 1px ${themeTokenRule("color-border-base")};

  /* stylelint-disable selector-max-type */
  tr:nth-child(even) {
    background-color: ${({ alternateRows }) =>
      alternateRows && themeTokenRule("color-background-alt")};
  }
`;

Table.displayName = "Table";
