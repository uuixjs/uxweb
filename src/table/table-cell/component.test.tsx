import { shallow } from "enzyme";
import { setupShallowTest } from "../../../tests/helpers";
import { TextAlign, VerticalAlign } from "../../layout";
import { TableCell, TableCellProps } from "./component";

function getOptionalProps(): TableCellProps {
  return {
    padding: 0.5,
    textAlign: TextAlign.Right,
    verticalAlign: VerticalAlign.TextTop,
  };
}

const setupShallow = setupShallowTest(TableCell);

describe("TableCell", () => {
  it("renders table cell component with required props", () => {
    const { wrapper } = setupShallow();
    expect(wrapper).toMatchSnapshot();
  });

  it("renders table cell component with all props", () => {
    const { wrapper } = setupShallow(getOptionalProps());
    expect(wrapper).toMatchSnapshot();
  });

  it("passes aria props to the element", () => {
    const wrapper = shallow(<TableCell role="tablecell" />);
    const element = wrapper.find("td").first();
    expect(element.prop("role")).toBe("tablecell");
  });
});
