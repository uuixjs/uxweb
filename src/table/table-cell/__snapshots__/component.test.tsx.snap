// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableCell renders table cell component with all props 1`] = `
<InjectLayout
  padding={0.5}
  textAlign="right"
  verticalAlign="text-top"
>
  <td
    className="tw-table-cell"
  />
</InjectLayout>
`;

exports[`TableCell renders table cell component with required props 1`] = `
<InjectLayout
  padding={1}
>
  <td
    className="tw-table-cell"
  />
</InjectLayout>
`;
